### PR TITLE
[SPARK-36101][CORE] Grouping exception in core/api

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
This PR group exception messages in core/src/main/scala/org/apache/spark/api

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce _any_ user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

**Comment by holdenk** [comment]
I'm not sure about the value here, it looks like it might make the code a bit harder for folks to grep around for error messages and it seems to be a net addition of lines making me think we're not really saving all that much repetition. Can you share a bit more about your thoughts on this?

**Comment by holdenk** [comment]
Jenkins ok to test

**Comment by mridulm** [comment]
I am find the method names a bit confusing - for example `juRemoveError`, `cannotBeUsedDataOfTypeError`.
While others like `taskInterruptionError`, `invalidPortNumberError` are clear (though should be `*Exception`).

But I agree with @holdenk, I dont see very compelling value proposition here.

**Comment by dgd-contributor** [comment]
cc @allisonwang-db 


> I'm not sure about the value here, it looks like it might make the code a bit harder for folks to grep around for error messages and it seems to be a net addition of lines making me think we're not really saving all that much repetition. Can you share a bit more about your thoughts on this?

This stage is only the first in the process. You can read [this](https://docs.google.com/document/d/1XGj1o3xAFh8BA7RCn3DtwIPC6--hIFOaNUNSlpaOIZs/edit#heading=h.c5ylwwsvjngo) for more details

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import java.io.EOFException
+
+import org.apache.spark.{SparkException, TaskKilledException}
+import org.apache.spark.api.r.JVMObjectId
+
+/**
+ * Object for grouping error messages from (most) exceptions thrown during query execution.
+ */
+object SparkCoreErrors {
+  def juRemoveError(): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:29]: removeUnsupportedError()

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    new UnsupportedOperationException("remove")
+  }
+
+  def unexpectedPy4JServerError(other: Object): Throwable = {
+    new RuntimeException(s"Unexpected Py4J server ${other.getClass}")
+  }
+
+  def cannotBeUsedDataOfTypeError(className: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:37]: dataOfTypeCannotBeUsedError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  def eofBeforePythonServerAcknowledgedError(): Throwable = {
+    new SparkException("EOF reached before Python server acknowledged")
+  }
+
+  def taskInterruptionError(reason: String): Throwable = {
+    new TaskKilledException(reason)
+  }
+
+  def crashedUnexpectedlyPythonWorkerError(eof: EOFException): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:57]: pythonWorkerCrashedError

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 4]
```diff
@@ -241,7 +240,7 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
             |  $pythonPath
             |Also, check if you have a sitecustomize.py module in your python path,
             |or in your python installation, that is printing to standard output"""
-          throw new SparkException(exceptionMessage.stripMargin)
+          throw SparkCoreErrors.invalidPortNumberError(exceptionMessage.stripMargin)
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala:243]: Shall we move the exception message inside the error function?

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+  def invalidPortNumberError(exceptionMessage: String): Throwable = {
+    new SparkException(exceptionMessage)
+  }
+
+  def failedToConnectBackPythonWorkerError(e: Exception): Throwable = {
+    new SparkException("Python worker failed to connect back.", e)
+  }
+
+  def eofExceptionWhileReadPortNumberFromAliveDaemonError(daemonModule: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:73]: eofExceptionWhileReadingPortNumberFromDaemonOutputError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+  def eofExceptionWhileReadPortNumberError(
+      daemonModule: String,
+      daemonExitValue: Int): Throwable = {
+    new SparkException(
+      s"EOFException occurred while reading the port number from $daemonModule's" +
+        s" stdout and terminated with code: $daemonExitValue.")
+  }
+
+  def cannotBeUsedRDDElementError(otherName: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:86]: RDDElementOfTypeCannotBeUsedError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    new SparkException(s"RDD element of type $otherName cannot be used")
+  }
+
+  def unsupportedDataTypeError(other: Any): Throwable = {
+    new SparkException(s"Data of type $other is not supported")
+  }
+
+  def workerProducedError(msg: String, e: Exception): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:94]: RWorkerExitedError

**Review by allisonwang-db — COMMENTED** [review]
Please also fix the merge conflicts. Thanks!

**Comment by dgd-contributor** [comment]
cc @allisonwang-db I have resolve what you suggested, can you re-review

**Review by allisonwang-db — APPROVED** [review]

**Comment by dgd-contributor** [comment]
cc @cloud-fan can you review

**Comment by dgd-contributor** [comment]
@cloud-fan hi, can you review this

**Comment by cloud-fan** [comment]
centralizing user-facing errors is a good direction (easy for auditing, support different languages), but we need to make sure they are user-facing errors.

There are a lot of user-facing errors in Spark SQL, but much less in Spark Core. There is no benefit to centralizing internal errors. We need to think deeper and make sure these errors are user-facing errors.

File: core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala [hunk 8]
```diff
@@ -89,7 +91,7 @@ private[spark] object JavaUtils {
                   mm.remove(k)
                   prev = None
                 case _ =>
-                  throw new UnsupportedOperationException("remove")
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala:92]: How can an end-user hit this error?

Comment (dgd-contributor) [reply core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala:92]: How can we identify which error is user-facing error or not? Is there any rule we can apply?

Comment (allisonwang-db) [reply core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala:92]: Usually, non-user-facing errors can be identified by the exception type, for example, `IllegalStateException`, or by the code logic. Here it seems if `underlying` is not a mutable map, then the `remove` operation is not supported. This seems to be a develop-facing error.

Comment (dgd-contributor) [reply core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala:92]: Thanks, so now we should determine which error is develop-facing error and restore it as before right?
What above previous pull requests that have already merged?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala:92]: > What above previous pull requests that have already merged?

Those should be revisited. It's not urgent as it's an internal refactor, and end-users won't notice it. We can enforce this in new PRs first.

**Comment by mridulm** [comment]
In general, atleast for core, unless it is a repeating pattern used atleast a few times and there is sufficient value in moving it out of the current throw site, let us not change it.

For example,I am more persuaded by changes to `Py4JServer` in this PR - while the changes to `PythonHadoopUtil`, `BaseRRunner`, `PythonRDD` and so on do not add value.

**Comment by dgd-contributor** [comment]
> In general, atleast for core, unless it is a repeating pattern used atleast a few times and there is sufficient value in moving it out of the current throw site, let us not change it.
> 
> For example,I am more persuaded by changes to `Py4JServer` in this PR - while the changes to `PythonHadoopUtil`, `BaseRRunner`, `PythonRDD` and so on do not add value.

cc @cloud-fan @allisonwang-db, would you refer this? Would I just keep the duplicated error and restore the other?

**Comment by cloud-fan** [comment]
> Would I just keep the duplicated error and restore the other?

We can go this way first. In the future let's come up with the error class first (which means the error is user-facing) and then move the error to the centralized file.

**Comment by dgd-contributor** [comment]
> > Would I just keep the duplicated error and restore the other?
> 
> We can go this way first. In the future let's come up with the error class first (which means the error is user-facing) and then move the error to the centralized file.

Can you re-review

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 9]
```diff
@@ -198,12 +213,12 @@ object SparkCoreErrors {
...
-  def failResourceOffersForBarrierStageError(errorMsg: String): Throwable = {
+  def sparkError(errorMsg: String): Throwable = {
     new SparkException(errorMsg)
   }
 
-  def markExecutorAsFailedError(errorMsg: String): Throwable = {
-    new SparkException(errorMsg)
+  def markExecutorAsFailedError(): Throwable = {
+    new SparkException("taskIdToTaskSetManager.contains(tid) <=> taskIdToExecutorId.contains(tid)")
   }
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:222]: Can we remove this method ? It is very specific to that codepath.

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 10]
```diff
@@ -31,6 +31,25 @@ import org.apache.spark.storage.{BlockId, BlockManagerId, BlockNotFoundException
...
+  def unexpectedPy4JServerError(other: Object): Throwable = {
+    new RuntimeException(s"Unexpected Py4J server ${other.getClass}")
+  }
+
+  def eofExceptionWhileReadPortNumberError(
+      daemonModule: String,
+      daemonExitValue: Option[Int] = null): Throwable = {
+    var msg = s"EOFException occurred while reading the port number from $daemonModule's" +
+      s" stdout"
+    if (daemonExitValue != null) {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:43]: Should be `daemonExitValue.isDefined`. Could be reformulated as
`val msg = s" ... stdout" + daemonExitValue.map(v => s" and terminated with code: $v.").getOrElse("")`

**Review by mridulm — APPROVED** [review]
+CC @cloud-fan Please take a look.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**